File: Assets/src/model/enemies/Hopper.cs
using System;
using System.Linq;
using UnityEngine;

[Serializable]
[ObjectInfo(description: "Jumps next to you.\nWhen hurt, it will eat a nearby Grass to heal itself to full HP.")]
public class Hopper : AIActor {
  public Hopper(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 6;
  }

  internal override (int, int) BaseAttackDamage() => (2, 2);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;

    if (hp != maxHp) {
      if (grass != null) {
        return new TelegraphedTask(this, 1, new GenericBaseAction(this, EatGrass));
      } else {
        var nearbyGrassTile = floor
          .BreadthFirstSearch(pos, t => t.CanBeOccupied() && DistanceTo(t) < 5)
          .Where(t => t.grass != null)
          .FirstOrDefault();
        if (nearbyGrassTile != null) {
          return new MoveToTargetTask(this, nearbyGrassTile.pos);
        } else {
          return new MoveRandomlyTask(this);
        }
      }
    } else {
      if (CanTargetPlayer()) {
        if (IsNextTo(player)) {
          return new AttackTask(this, player);
        } else {
          var jumpTile = floor.GetAdjacentTiles(player.pos).Where(t => t.CanBeOccupied()).OrderBy(t => t.DistanceTo(pos)).FirstOrDefault();
          return new JumpToTargetTask(this, jumpTile.pos);
        }
      } else {
        return new MoveRandomlyTask(this);
      }
    }
  }

  void EatGrass() {
    if (grass != null) {
      Heal(maxHp - hp);
      grass.Kill(this);
    }
  }
}

[Serializable]
class JumpToTargetTask : DoOnceTask {
  public JumpToTargetTask(Actor actor, Vector2Int position) : base(actor) {
    this.position = position;
  }

  public Vector2Int position { get; }

  protected override BaseAction GetNextActionImpl() {
    return new JumpBaseAction(actor, position);
  }
}

public class JumpBaseAction : BaseAction {
  public override ActionType Type => ActionType.MOVE;
  public readonly Vector2Int pos;

  public JumpBaseAction(Actor actor, Vector2Int pos) : base(actor) {
    this.pos = pos;
  }

  public override void Perform() {
    actor.pos = pos;
  }
}

File: web/src/model/enemies/Hopper.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { MoveToTargetTask } from '../tasks/MoveToTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { JumpToTargetTask } from '../tasks/JumpToTargetTask';
import { TelegraphedTask } from '../tasks/TelegraphedTask';
import { WaitTask } from '../tasks/WaitTask';
import { GenericBaseAction } from '../BaseAction';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';

/**
 * Jumps next to you. When hurt, eats a nearby Grass to heal to full HP.
 * Port of C# Hopper.cs.
 */
export class Hopper extends AIActor {
  constructor(pos: Vector2Int) {
    super(pos);
    this._hp = 3;
    this._baseMaxHp = 3;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;

    if (this._hp !== this.maxHp) {
      // Hurt — try to eat grass
      if (this.grass != null) {
        return new TelegraphedTask(this, 1, new GenericBaseAction(this, () => this.eatGrass()));
      } else {
        // Search for nearby grass
        const nearbyGrassTile = this.floor!
          .breadthFirstSearch(this.pos, t => t.canBeOccupied() && this.distanceTo(t.pos) < 5)
          .find(t => t.grass != null);
        if (nearbyGrassTile) {
          return new MoveToTargetTask(this, nearbyGrassTile.pos);
        } else {
          return new MoveRandomlyTask(this);
        }
      }
    } else {
      // Full HP — chase and attack
      if (this.canTargetPlayer()) {
        if (this.isNextTo(player)) {
          return new AttackTask(this, player);
        } else {
          const jumpTile = this.floor!
            .getAdjacentTiles(player.pos)
            .filter(t => t.canBeOccupied())
            .sort((a, b) => a.distanceTo(this.pos) - b.distanceTo(this.pos))
            [0];
          if (jumpTile) {
            return new JumpToTargetTask(this, jumpTile.pos);
          } else {
            return new WaitTask(this, 1);
          }
        }
      } else {
        return new MoveRandomlyTask(this);
      }
    }
  }

  private eatGrass(): void {
    if (this.grass != null) {
      this.heal(this.maxHp - this._hp);
      this.grass.kill(this);
    }
  }
}

entityRegistry.register('Hopper', Hopper);
